### PR TITLE
[WIP]TLG decoding and 4-gram lattice rescore

### PR DESCRIPTION
**Pull request by glynpu**
As discussed in https://github.com/espnet/espnet/pull/3358, this pr is about TLG decoding and 4-gram lattice rescore in following whole plan:

0. encoder + ctc decocer （#3358）
1.  support batch decoding (#3433)
2.  support ctc_topo + 3-gram decoding, which will be referred as TLG decoding and T is short for "Token". (_**current pr**_)
3.  4-gram lattice rescore (**_current pr_**)
4.  transformer-decoder nbest rescore
5.  neural net language model nbest rescore

### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2/test_clean|2620|52576|97.4|2.4|0.2|0.3|3.0|34.9|
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_use_k2_tlg_fgram_rescoring/test_clean|2620|52576|97.1|2.2|0.7|0.3|3.2|37.0|
### CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2/test_clean|2620|281530|99.3|0.3|0.3|0.4|1.0|34.9|
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_use_k2_tlg_fgram_rescoring/test_clean|2620|281530|98.6|0.3|1.1|0.2|1.6|37.0|

>Can you add the above decoding time results to README.md to record the effectiveness of batch decoding in the next PR?

@sw005320 The decoding time will be seriously measured and added in README.md after this pr is reviewed. 
An unknown **_bug_** seems exists since wer on test-clean of (TLG + 4-gram lattice resocre) is **_3.2%_** while that of ctc decoding is **_3.0%_**.

### CODE REVIEW

**Comment by alucassch** [comment]
I have a large phone aligned dataset where the text has the correct pronounce for every word in all utterances. I trained a model with espnet with bpe subwords of this text. As in this case the model gives me phones in the output I am trying to use k2 to give me back the corresponding words. To do this I used the bpe model to build a lexicon where each word has many phonetic pronounces but every pronounce is represented with the corresponding pieces given by the bpe model.

However during decoding I have some words that are completely deleted even when I know the CTC outputs the correct tokens. There is no pattern of deletion, something similar to what was related here https://github.com/k2-fsa/snowfall/pull/230 with disambig symbols

I was writing this pipeline before seeing this PR and face this issue. Using this PR and changing the `prepare_dict.sh` to write the correct lexicon.txt file gives me the exact same errors. Any idea or sugestion?

**Comment by glynpu** [comment]
> I was writing this pipeline before seeing this PR and face this issue. Using this PR and changing the `prepare_dict.sh` to write the correct lexicon.txt file gives me the exact same errors. Any idea or sugestion?

@alucassch Is it possible to release your code and trained model? It's easier to diagnose with that.

**Comment by danpovey** [comment]
@alucassch I suspect the problem has to do with a stage such as removing epsilons that generated a ragged attribute for the words (with homophones, some word symbols can end up at the start of the next word due to disambig symbols being removed)... and somehow, later on, these ragged attributes are not treated correctly when extracting the text, but I forget the exact mechanism for this.  I seem to remember that this was related to a change in the behavior of the k2 library that happened at some point, however, again, I don't remember the details of this, likely seeing your code will trigger a memory.

**Comment by danpovey** [comment]
... regardless, using an up to date version of k2 may be helpful.

**Comment by glynpu** [comment]
>An unknown bug seems exists since wer on test-clean of (TLG + 4-gram lattice resocre) is 3.2% while that of ctc decoding is 3.0%.

Tuning lm_scores weight gives a better result, now wer on test-clean of (TLG + 4-gram lattice rescore) decreases from previously 3.2%(lm_weight = 1.0) to 2.7%(lm_weight=0.5). 

In short, wer of ctc_decoding is 3.0% and of (TLG + 4-gram lattice rescore) is 2.7%. 
Following are detail errors:

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2/test_clean|2620|52576|97.4|2.4|0.2|0.3|3.0|34.9|
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_use_k2_tlg_fgram_rescoring/test_clean|2620|52576|97.6|2.1|0.2|0.3|2.7|32.2|

**Comment by sw005320** [comment]
Can you paste the performance for each lm weight here for a record?

File: egs2/TEMPLATE/asr1/asr.sh [hunk 1]
```diff
@@ -94,6 +94,16 @@ num_splits_asr=1           # Number of splitting for lm corpus.
 
 # Decoding related
 use_k2=false      # Whether to use k2 based decoder
+is_ctc_decoding=false
+inference_with_tlg=false
+if $inference_with_tlg && $is_ctc_decoding ; then
+  echo "Please chose ctc decoding or tlg decoding."
+  echo "Either of them should be false."
+fi
+lang_dir=data/lang_bpe/
```
Comment (sw005320) [inline egs2/TEMPLATE/asr1/asr.sh:103]: `lang_dir` and `lang_bpe` are not so easy to understand the meaning.
Can you change it more appropriate ones?

My understanding is that this directory is used to store LMs (n-grams) and FSTs.
We may use fst as a keyword, e.g., lang_fst_dir and data/lang_fst

**Comment by glynpu** [comment]
As discussed in #3358, the final whole pipeline is use (encoder + TLG decoding  + 4gram lattice rescore + transformer decoder nbest rescore + transformer lm nbest rescore).

Now following three configuration is implemented. 
I am still figuring out a way to combine 4-gram lattice rescoring with transformer-decoder and transformer lm nbest rescoring.
Currently, The result of directly combing lattice rescoring part and nbest-rescoring part is not good. 

The acceptable results are listed below:

 test-clean wer | encoder + k2-ctc-decode | encoder + k2 ctc decode +3-gram+4-gram | encoder + ctc decode + 6-layer-decoder nbest rescore + transformer lm nbest rescore| encoder + k2 ctc decode +3-gram+4-gram+ 6-layer-decoder nbest rescore + transformer lm nbest rescore
-- | -- | -- | -- | --
baseline: result with espnet decoding method | * |  *| 2.1% decoder algorithm in espnet|*
 decode with k2| 3.0 | 2.7  | 2.3|  to be implemented


with following details errors:
|method|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|ctc_decoding|2620|52576|97.4|2.4|0.2|0.3|3.0|34.9|
|TLG decoding (ctc + 3-gram) + 4-gram lattice rescore|2620|52576|97.6|2.1|0.2|0.3|2.7|32.2|
|ctc_decoding + nbest rescoring with transcormer-decoder and transformer language model(n=1000)|2620|52576|98.0|1.8|0.2|0.3|2.3|27.5|

File: egs2/TEMPLATE/asr1/asr.sh [hunk 2]
```diff
@@ -409,6 +419,15 @@ if [ -z "${inference_tag}" ]; then
...
     if "${use_k2}"; then
       inference_tag+="_use_k2"
+      if "${inference_with_tlg}"; then
+        inference_tag+="_use_k2_tlg"
+        if ${use_fgram_rescoring}; then
+          inference_tag+="_fgram_rescoring"
+        fi
+      fi
+      if ${use_nbest_rescoring}; then
+        inference_tag+="_nbest_rescoring"
```
Comment (sw005320) [inline egs2/TEMPLATE/asr1/asr.sh:429]: Is it nbest restoring?
I thought that it is lattice rescoring

File: egs2/librispeech/asr1/local/fstaddselfloops.pl [hunk 3]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env perl
```
Comment (sw005320) [inline egs2/librispeech/asr1/local/fstaddselfloops.pl:1]: This looks a general fst operation. We may move it to https://github.com/espnet/espnet/tree/master/egs2/TEMPLATE/asr1/scripts In the future.

File: egs2/librispeech/asr1/local/make_lexicon_fst.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
+#!/usr/bin/env python3
```
Comment (sw005320) [inline egs2/librispeech/asr1/local/make_lexicon_fst.py:1]: Ditto.
We can move to https://github.com/espnet/espnet/tree/master/egs2/TEMPLATE/asr1/scripts In the future.

File: egs2/TEMPLATE/asr1/asr.sh (hunk 1 above)
Comment (sw005320) [reply egs2/TEMPLATE/asr1/asr.sh:103]: If there is some reasons (k2 conventions and so on), we can keep it.
Please let me know your opinions.

File: egs2/librispeech/asr1/local/prepare_dict.sh [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      $lang_dir/lexiconp_disambig.txt | \
+      utils/sym2int.pl --map-oov 1 -f 3 $lang_dir/tokens.txt | \
+      utils/sym2int.pl -f 4 $lang_dir/words.txt  | \
+      local/fstaddselfloops.pl $wdisambig_token $wdisambig_word > $lang_dir/L_disambig.fst.txt || exit 1
+  fi
+
+fi
+
+if [ $stage -le 5 ]; then if [ ! -f $lang_dir/G.fst.txt ]; then
+    python3 -m kaldilm \
```
Comment (sw005320) [inline egs2/librispeech/asr1/local/prepare_dict.sh:92]: I think you need to install kaldilm (or is it already included in the k2 installation)?

**Comment by glynpu** [comment]
>Can you paste the performance for each lm weight here for a record?

Sure.
 Actually I am struggling with combining with 4-gram lattice rescore and (transformer decoder + transformer lm) nbest rescore  part. 
I am implementing a weight-search module which maybe both useful to get performance of different lm weight as you mentioned also to search an ideal weight combination of 4-gram_lm_score, decoder_score and nnlm_score.

File: espnet2/bin/k2_asr_inference.py [hunk 6]
```diff
@@ -1,6 +1,8 @@
 #!/usr/bin/env python3
```
Comment (sw005320) [inline espnet2/bin/k2_asr_inference.py:1]: I forgot to mention it in the previous PR.
I think it’s better to rename it `asr_inference_k2.py`.
Same prefix of the function name makes extensions easier.

E.g., If we have other inference engine (e.g., k3), we may add asr_inference_k3.py, and people can easily understand that k3 is a variant of the decoding method.
If we extend k2 to add some new functions, e.g., say k2_fast, then asr_inference_k2_fast.py.

File: espnet2/bin/k2_asr_inference.py [hunk 7]
```diff
@@ -221,25 +294,138 @@ def __call__(
...
+        if self.use_fgram_rescoring:
+            lattices = rescore_with_whole_lattice(
+                lattices, self.G, lm_scale_list=None, need_rescored_lats=True
+            )
 
-        best_paths = k2.shortest_path(lattices, use_double_scores=True)
-        scores = best_paths.get_tot_scores(
-            use_double_scores=True, log_semiring=False
-        ).tolist()
+        results = []
```
Comment (sw005320) [inline espnet2/bin/k2_asr_inference.py:302]: I’ll check this part later.

File: espnet2/fst/common.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
+import logging
```
Comment (sw005320) [inline espnet2/fst/common.py:1]: @takaaki-hori, can you check this part?

File: espnet2/bin/k2_asr_inference.py (hunk 7 above)
Comment (sw005320) [reply espnet2/bin/k2_asr_inference.py:302]: @takaaki-hori, can you check this part?

File: egs2/TEMPLATE/asr1/asr.sh (hunk 2 above)
Comment (sw005320) [reply egs2/TEMPLATE/asr1/asr.sh:429]: Sorry.
I think you’re preparing this for neural LMs or attention encoder-decoder rescoring, right?
If so, could you add some comments about it?

File: espnet2/fst/lm_rescore.py [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+            new_tokens.append(token)
+            previous = token
+    new_tokens = [token for token in new_tokens if token > blank_id]
+    return new_tokens
+
+
+def compute_am_scores_and_fm_scores(
+    lats: k2.Fsa, word_fsas_with_epsilon_loops: k2.Fsa, path_to_seq_map: torch.Tensor
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Compute AM and LM scores of n-best lists (represented as word_fsas).
```
Comment (sw005320) [inline espnet2/fst/lm_rescore.py:37]: I think the AM score is not correct terminology.
AM should be replaced with e2eASR or something like that?

Comment (sw005320) [reply espnet2/fst/lm_rescore.py:37]: If it requires too many changes or not aligned with k2’s terminology or we only limit this use for CTC, of course you can keep “am”. 
But I think it’s better to add a note about this terminology.

File: espnet2/fst/lm_rescore.py [hunk 10]
```diff
@@ -0,0 +1,569 @@
...
+    previous = None
+    for token in tokens:
+        if token != previous:
+            new_tokens.append(token)
+            previous = token
+    new_tokens = [token for token in new_tokens if token > blank_id]
+    return new_tokens
+
+
+def compute_am_scores_and_fm_scores(
```
Comment (sw005320) [inline espnet2/fst/lm_rescore.py:34]: Why fm_scores instead of lm_scores?

Comment (sw005320) [inline espnet2/fst/lm_rescore.py:34]: There are a couple of duplicated functions named `compute_am_scores_and_fm_scores`

**Review by sw005320 — COMMENTED** [review]
Many thanks!
I’m reviewing your PR.
(A bit shallow review)

Do we need to install openfst?
If so, we may need to prepare the installation and corresponding document.
If not, this is great! Then, it is still better to emphasize it in the document.

**Comment by alucassch** [comment]
> @alucassch Is it possible to release your code and trained model? It's easier to diagnose with that.

I'm working on a model that I can share publicly. It will be finished soon. I also tried update k2 to latest (as stated by @danpovey ) but got the same results. Meanwhile I will check if these new updates with rescoring give me any different results.

**Comment by glynpu** [comment]
This commit is mainly for cleaning up the code.

There are mainly two branches to do the integration of k2. 
One is 70d16d210cdce28e71b7892a0ec96eaaa6474d64(Update librispeech result), which maybe refered as  _**wer_computation_branch**_. 
The other one is based on master, which maybe refered as "**_master_based_branch_**". 
Because the released model is only compatible with the "**_wer_computation_branch_**".  New implemented parts will be first tested and wer is computated on that branch and then code will be copied to the "**_master_based_branch_**" with git stash.
Somehow, I made some mistake in the copying process,  so the code is totally messed up and the ci also fails.

Anyway, the code now is what it should be. You may spare sometime to help me review it. @sw005320 

I will fix the reviews you alread given soon.  

Here is the details result with current code.  There seems something wrong with TLG decoding with 3.2(row 3) > 3.0 (row 2) and 2.5(row 5) > 2.3(row 4)

### WER
|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|CTC_decoding/test_clean|2620|52576|97.4|2.4|0.2|0.3|3.0|34.8|
|TLG decoding/test_clean|2620|52576|97.1|2.3|0.7|0.3|3.2|38.1|
|CTC_decoding+nbest_rescoring/test_clean|2620|52576|98.0|1.8|0.2|0.3|2.3|27.7|
|TLG decoding + nbest_rescoring/test_clean|2620|52576|97.8|1.9|0.3|0.3|2.5|30.1|
|TLG decoding +fgram_rescoring/test_clean|2620|52576|97.6|2.1|0.2|0.3|2.7|32.0|

**Comment by sw005320** [comment]
Let's restart this PR.
In addition to my comments, could you add more tests to increase the test coverage?

**Comment by glynpu** [comment]
> Let's restart this pr.
> In addition to my comments, could you add more tests to increase the test coverage?

@sw005320 Could I make another PR which only contains nbest rescoring pipeline(which got 2.3% on test_clean in above table), i.e., remove TLG and 4-gram part based on this pr. 

For the following reasons:
1. TLG decoding pipeline may need more time than expected. 
    It turns out that the released model used in thi pr triggers a bug in k2 when I tune TLG decoding, which seems caused by a third-dependent library.  It may cost some time to be fixed. And before that, it make no sense to tune this TLG related decoding pipeline(which got 2.5% on test_clean).
2. nbest decoding pipeline(which got 2.3% on test-clean) is ready for merge and easy to review.
    Now it is mixed toghther with TLG decoding pipeline, making this pr quite heavy.

**Comment by sw005320** [comment]
> > Let's restart this pr.
> > In addition to my comments, could you add more tests to increase the test coverage?
> 
> @sw005320 Could I make another PR which only contains nbest rescoring pipeline(which got 2.3% on test_clean in above table), i.e., remove TLG and 4-gram part based on this or.

Yes.
I actually prefer to have smaller PR.
Let's do it!!!

**Pull request closed without merging**